bundles/org.openhab.persistence.mysql/src/main/java/org/openhab/persistence/mysql/internal/MysqlPersistenceService.java

```diff
@@ -0,0 +1,717 @@
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.OpenClosedType;
+import org.openhab.core.library.types.PercentType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.persistence.FilterCriteria;
+import org.openhab.core.persistence.FilterCriteria.Ordering;
+import org.openhab.core.persistence.HistoricItem;
+import org.openhab.core.persistence.PersistenceItemInfo;
+import org.openhab.core.persistence.PersistenceService;
+import org.openhab.core.persistence.QueryablePersistenceService;
+import org.openhab.core.persistence.strategy.PersistenceStrategy;
+import org.openhab.core.types.State;
+import org.openhab.core.types.UnDefType;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is the implementation of the mySQL {@link PersistenceService}.
+ *
+ * Data is persisted with the following conversions:
+ *
+ * Item-Type Data-Type MySQL-Type
+ * ========= ========= ==========
+ * ColorItem HSBType CHAR(25)
+ * ContactItem OnOffType CHAR(6)
+ * DateTimeItem DateTimeType DATETIME
+ * DimmerItem PercentType TINYINT
+ * NumberItem DecimalType DOUBLE
+ * RollershutterItem PercentType TINYINT
+ * StringItem StringType VARCHAR(20000)
+ * SwitchItem OnOffType CHAR(3)
+ *
+ * In the store method, type conversion is performed where the default type for
+ * an item is not as above. For example, DimmerType can return OnOffType, so to
+ * keep the best resolution, we store as a number in SQL and convert to
+ * DecimalType before persisting to MySQL.
+ *
+ * @author Henrik Sjöstrand - Initial contribution
+ * @author Thomas.Eichstaedt-Engelen - Enhancements
+ * @author Chris Jackson - Enhancements
+ * @author Helmut Lehmeyer - Enhancements
+ */
+@NonNullByDefault
+@Component(service = { PersistenceService.class,
+        QueryablePersistenceService.class }, configurationPid = "org.openhab.mysql", configurationPolicy = ConfigurationPolicy.REQUIRE)
+public class MysqlPersistenceService implements QueryablePersistenceService {
+
+    private static final Pattern EXTRACT_CONFIG_PATTERN = Pattern.compile("^(.*?)\\.([0-9.a-zA-Z]+)$");
+
+    private static final Logger logger = LoggerFactory.getLogger(MysqlPersistenceService.class);
+
+    private String driverClass = "com.mysql.jdbc.Driver";
+    private @NonNullByDefault({}) String url;
+    private @NonNullByDefault({}) String user;
+    private @NonNullByDefault({}) String password;
+
+    private boolean initialized = false;
+
+    @Reference
+    protected @NonNullByDefault({}) ItemRegistry itemRegistry;
+
+    // Error counter - used to reconnect to database on error
+    private int errCnt;
```

Comment:
I don't know what kind of multithreading would go on in this class, but for threadsafety I would change this to AtomicInteger